src/GraphQL.Tests/Types/AutoRegisteringObjectGraphTypeTests.cs

```diff
@@ -363,6 +363,39 @@ public void FieldResolversWork(string fieldName, object expected)
             actual.ShouldBe(expected);
         }
 
+        [Fact]
+        public void WorksWithNoDefaultConstructor()
+        {
+            var graphType = new TestFieldSupport<NoDefaultConstructorTest>();
+            var context = new ResolveFieldContext
+            {
+                Source = new NoDefaultConstructorTest(true)
+            };
+            graphType.Fields.Find("Example1")!.Resolver!.Resolve(context).ShouldBe(true);
+            graphType.Fields.Find("Example2")!.Resolver!.Resolve(context).ShouldBe("test");
+            graphType.Fields.Find("Example3")!.Resolver!.Resolve(context).ShouldBe(1);
+        }
+
+        [Fact]
+        public void ErrorWithNoDefaultConstructorWhenSourceNull()
+        {
+            var graphType = new TestFieldSupport<NoDefaultConstructorTest>();
+            var context = new ResolveFieldContext();
+            Should.Throw<NullReferenceException>(() => graphType.Fields.Find("Example1")!.Resolver!.Resolve(context).ShouldBe(true));
```

Comment:
NRE?

Reply:
Could this sort of type without default ctor be instantiated via DI ?

Reply:
> DI

As it is with this PR, you can override the BuildSourceExpression with `context => ActivatorUtilities.GetServiceOrCreateInstance<TSourceType>(context.RequestServices)` to always pull from DI and similar. Although that specific method is only with the MS DI package.

Reply:
I have added a sample of how a derived class may use the `AutoRegisteringObjectGraphType` for dependency injection.  I have code set up similar to this in some of my projects, and I would like the test to remain here, not so much as a sample, but so as to verify that the desired functionality can be achieved in a derived class.  The derived class makes use of almost everything that has been added to `AutoRegisteringObjectGraphType` so far.  We should probably make some of the `AutoRegisteringHelper` methods public to facilitate use such as is demonstrated.  I'll probably write a PR for that after other modifications are complete.

Reply:
Just FYI, I think that the functionality within `AutoRegisteringObjectGraphType`, once we add this PR, should be sufficient for most use cases.  While it is possible to go the next step with DI injection as demonstrated in the sample included within this PR, I don't think that will really be necessary to include as a built-in feature.  It's certainly worth discussing of course, but now that we have `[FromServices]` available, it should be really easy to inject services where needed, and has the benefit of only using the required service for the field that requires it -- of course that's at the cost of not having the services listed in the constructor.

Reply:
> Making this type of object the root of a graph would result in NRE.

I don't understand.

Reply:
> I don't understand.

So historically, the point of AutoRegisteringObjectGraphType is this:

```gql
{
  people    # Query graph returns IEnumerable<Person>
  {         # this graph type is AutoRegisteringObjectGraphType<Person>
    id      # returns ((Person)IResolveFieldContext.Source).Id
    name    # returns ((Person)IResolveFieldContext.Source).Name
  }
}
```

Now we have added the capability to write methods, and even those with arguments, and change names via attributes and so on:

```gql
{
  people                  # Query graph returns IEnumerable<Person>
  {                       # this graph type is AutoRegisteringObjectGraphType<Person>
    id                    # returns ((Person)IResolveFieldContext.Source).Id
    name                  # returns ((Person)IResolveFieldContext.Source).Name
    children(minAge: 5)   # returns ((Person)IResolveFieldContext.Source).GetChildren(5)
    {
      id
      name
    }
  }
}
```

Now normally if this was used as the query root it would fail unless `ExecutionOptions.Root` was set

```gql
{                # schema.Query = typeof(AutoRegisteringObjectGraphType<Person>);
  id             # NRE trying to access IResolveFieldContext.Source
}
```

However, this is a useful scenario if you just want to easily write some methods on a class, for example let's go back to the original query:

```gql
{
  people
  {
    ...
  }
}
```

If you want to write this with C# code like this:

```cs
public class Query
{
    public IEnumerable<Person> People([FromServices] db) => db.GetPeople();
}
```

This is going to be a problem when the query executes:

```gql
{
  people      # returns ((Query)IResolveFieldContext.Source).People()    -- which is a NRE because Source is null
  {
    ...
  }
}
```

Obviously we could set the root to `new Query()`.  But then if we wrote a similar bit of code for Mutation, we would end up with `InvalidCastException`s because root is `new Query()` instead of `new Mutation()`.

Now if we think about it, when is `IResolveFieldContext.Source` null?  It can ONLY ever be in the above situation.  If null were returned for any other point in the graph, the execution strategy would not execute child nodes.  So `Source` always contains a value, except when used as the `Query`, `Mutation` or `Subscription` types, and only when `Root` is null.

This PR changes the resolver like this (going to the first example):

```gql
{
  people    # Query graph returns IEnumerable<Person>
  {         # this graph type is AutoRegisteringObjectGraphType<Person>
    id      # returns (((Person)IResolveFieldContext.Source) ?? new Person()).Id
    name    # returns (((Person)IResolveFieldContext.Source) ?? new Person()).Name
  }
}
```

Normally nothing is different.  The null check should not add any significant time to execution and no allocations are added.  However now we can use `AutoRegisteringObjectGraphType` as in the example above:

```gql
{
  people      # returns (((Query)IResolveFieldContext.Source) ?? new Query()).People()
  {
    ...
  }
}
```

Because now Query is instantiated.  It works with `ScopedAttribute` and stuff too if needed.

However, a NRE is still possible if `Query` does not have a default constructor, because then it will build the field resolver without the null-coalescing operator (just as it always did before, even before ANY of my changes).  I wrote a test for this condition.  I certainly could add code so that something like this runs, if you think I should:

```gql
{
  people      # returns (((Query)IResolveFieldContext.Source) ?? throw new InvalidOperationException($"Cannot use type 'Query' as a root graph type because no default constructor was found.")).People()
  {
    ...
  }
}
```

Reply:
Now another solution to the problem as a whole is to support static methods.  So then you would write this:

```cs
schema.Query = typeof(AutoRegisteringGraphType<Query>);

public class Query
{
    public static IEnumerable<Person> People([FromServices] db) => db.GetPeople();
}
```

Then it is not necessary to use null-coalescing operators within the field resolvers.  It would be worthwhile at this point to add `[FromSource]` into the codebase also.

Reply:
See:
- #2938

Reply:
Merged #2938 - since you approved it, I assume you prefer that option.

Reply:
So now it does not implicitly create objects, but it does support static methods, and that is how you would create a graph from a clr type for a root object.

Reply:
OK